### PR TITLE
initial commit for test

### PR DESCRIPTION
**Pull request by sibicramesh**

### CODE REVIEW

File: enforcer/utils/packetgen/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package packetgen
+
+import (
+	"net"
+
+	"github.com/google/gopacket/layers"
+)
+
+type Pkt interface {
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Dont call it Pkt here and Packet below.

File: enforcer/utils/packetgen/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	GenerateIPPacket(srcIPstr string, dstIPstr string) layers.IPv4
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Should return PacketIf ?

File: enforcer/utils/packetgen/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	GenerateTCPPacket(ip *layers.IPv4, srcPort layers.TCPPort, dstPort layers.TCPPort) layers.TCP
+	ChangeSequenceNumber(seqNum uint32) layers.TCP
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Why is it returning anything ?

Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Should have TCP in function name. Remove ambiguity.

File: enforcer/utils/packetgen/interfaces.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	ChangeAcknowledgementNumber(ackNum uint32) layers.TCP
+	ChangeWindow(window uint16) layers.TCP
+	SetSynTrue() layers.TCP
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: True is not needed. SetTCPSynFlag is good.

File: enforcer/utils/packetgen/interfaces.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	SetSynAckTrue() layers.TCP
+	SetAckTrue() layers.TCP
+	ChangePayload(newPayload string) layers.TCP
+	GetChecksum() uint16
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: IP Checksum or TCP ?

File: enforcer/utils/packetgen/interfaces.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+type Pkt interface {
+	GenerateIPPacket(srcIPstr string, dstIPstr string) layers.IPv4
+	GenerateTCPPacket(ip *layers.IPv4, srcPort layers.TCPPort, dstPort layers.TCPPort) layers.TCP
+	ChangeSequenceNumber(seqNum uint32) layers.TCP
+	ChangeAcknowledgementNumber(ackNum uint32) layers.TCP
+	ChangeWindow(window uint16) layers.TCP
+	SetSynTrue() layers.TCP
+	SetSynAckTrue() layers.TCP
+	SetAckTrue() layers.TCP
+	ChangePayload(newPayload string) layers.TCP
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Is this Appending if a payload already exists ? 
 - AppendTCPPayload() - what is the return type and why ?
Does it error if a payload exists ?
- NewTCPPayload() and should return error

File: enforcer/utils/packetgen/interfaces.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	GetChecksum() uint16
+}
+
+type PktFlow interface {
+	GenerateTCPFlow(bytePacket [][]byte) [][]byte
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Fix return types.

File: enforcer/utils/packetgen/interfaces.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	//GenerateInvalidTCPFlow() [][]byte
+	GetSynPackets() [][]byte
+	GetSynAckPackets() [][]byte
+	GetAckPackets() [][]byte
+}
+
+type Packet struct {
+	IPLayer            layers.IPv4
+	TCPLayer           layers.TCP
+	SequenceNum        uint32
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Delete from here till line 40.

File: enforcer/utils/packetgen/interfaces.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	Window             uint16
+	SrcIPstr, DstIPstr string
+	SrcIP, DstIP       net.IP
+	SrcPort, DstPort   layers.TCPPort
+	Layers             [150]layers.TCP
+}
+
+type PacketFlow struct {
+	Packet
+	GeneratedFlow [][]byte
```
Comment (satyamsi) [inline enforcer/utils/packetgen/interfaces.go]: Flow only.  TemplateFlow is another object.

File: enforcer/utils/packetgen/packet_gen_test.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"testing"
+)
+
+//. "github.com/smartystreets/goconvey/convey"
+
+var (
+	TCPFlow        [][]byte
+	InvalidTCPFlow [][]byte
+	layer          PacketFlow
```
Comment (satyamsi) [inline enforcer/utils/packetgen/packet_gen_test.go]: rename layer and fix types.

File: enforcer/utils/packetgen/packet_gen.go [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+//NewPacket returns a packet strut which implements PacketManipulator
+func NewPacket() PacketManipulator {
+
+	return &Packet{}
+
```
Comment (satyamsi) [inline enforcer/utils/packetgen/packet_gen.go]: Be consistent. Remove these blanks at trailing edge.

File: enforcer/utils/packetgen/packet_gen.go [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+//SetTCPSequenceNumber changes TCP sequence number
+func (p *Packet) SetTCPSequenceNumber(seqNum uint32) error {
+
+	if p.TCPLayer.Seq != 0 {
+		return fmt.Errorf("Sequence number already exists")
+	}
+
+	p.TCPLayer.Seq = seqNum
+
```
Comment (satyamsi) [inline enforcer/utils/packetgen/packet_gen.go]: How do you plan to handle checksum fixes. Do you want one to do like 

SetTCPSequenceNumber
 and then
ComputeTCPChecksum

File: enforcer/utils/packetgen/packet_gen.go [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+	return *p.TCPLayer
+
+}
+
+//ToBytes creates a packet buffer and converts it into a complete packet with ethernet, IP and TCP (with options)
+func (p *Packet) ToBytes() [][]byte {
+
+	//Creating a ethernet packet with fixed MAC addresses
+	ethernetLayer := layers.Ethernet{
```
Comment (satyamsi) [inline enforcer/utils/packetgen/packet_gen.go]: This (EtheretHeader) should exist in the packet and allowed to be manipulated. Not here.

File: enforcer/utils/packetgen/packet_gen.go (hunk 12 above)
Comment (sibicramesh) [inline enforcer/utils/packetgen/packet_gen.go]: So the gopacket has a feature to compute checksum on the fly. After changing Sequence Number or adding NewPayload, While serializing the layers, It computes the checksum.

Reference: https://github.com/google/gopacket/blob/master/layers/tcpip.go

Correct me If I am wrong!

Comment (satyamsi) [inline enforcer/utils/packetgen/packet_gen.go]: Thats good. So make sure that when someone asks for GetTCPPacket or GetTCPChecksum or ToBytes gets called the checksum is updated.

**Pull request closed without merging**